File: 07.py
import collections, re, networkx as nx

TEST=0#1
lines = open(0).read().splitlines()
D = collections.defaultdict(int)
g = nx.DiGraph() # p2
for line in lines:
    names = re.findall(r'[a-z]+', line)
    w = int(re.findall(r'\((\d+)\)', line)[0])
    for name in names: D[name] += 1
    # p2
    g.add_node(names[0], w=w)
    if '->' in line:
        for i in range(1, len(names)):
            g.add_edge(names[0], names[i])

p1 = None
for k,v in D.items():
    if v == 1:
        p1 = k
        break

p2 = None
if TEST:
    for node in g:
        print('node/',node)
        if g[node]: print('\t↳',','.join(g[node]))
    for it in g.nodes.items(): print('item/',it)
    for ig in g.in_degree(): print('in-degree/',ig)
    for og in g.out_degree(): print('out-degree/',og)
    for dg in g.degree: print('degree/',dg)
ID = dict(g.in_degree())
OD = dict(g.out_degree())

rev = list(nx.topological_sort(g))[::-1]
SUMS = {}
# traverse upwards in a reversed tree, one level at a time
for node in rev:
    if TEST:
        print('node -',node,g.nodes[node]['w'],'in/out',ID[node],OD[node])
    children = g[node]
    FREQ = collections.defaultdict(int)
    for child in children:
        FREQ[SUMS[child]] += 1
    N = len(FREQ)
    special = None
    assert N in range(3)
    if N == 2:
        many = max(FREQ, key=FREQ.get)
        only = min(FREQ, key=FREQ.get)
    update = 0
    for child in children:
        update += SUMS[child]
        accwei = SUMS[child]
        if N == 2 and accwei != many:
            special = child
            break
    if special:
        correction = abs(many - only)
        p2 = g.nodes[special]['w'] - correction
        print(f'disc/ {list(FREQ.items())}')
        print(f'found/ {special}\ncorrection/ {correction}')
        break
    SUMS[node] = update + g.nodes[node]['w']

print('part 1:', p1)
print('part 2:', p2)
assert p1 == list( nx.topological_sort(g) )[0]
assert p1 in ['gynfwly','tknk']
assert p2 in [60,1526]

File: 07.cc
#include "iostream"
#include "vector"
#include "unordered_map"
#include "sstream"
#include "deque"

struct Node
{ int weight = 0; std::vector<std::string> progs; };

struct defaultdict
{
    std::unordered_map<std::string, Node> _map;
    Node & operator[] (const std::string & key)
    { return _map[key]; }
};

template <typename ... Params>
void print(const Params &... args)
{ ((std::cout << args << ' '), ...) << std::endl; }

std::string join(const std::vector<std::string> & v)
{
    std::string res = join(v);
    res.pop_back();
    return res;
}

int main() {
    defaultdict g;
    std::string line;
    while (std::getline(std::cin, line)) {
        std::stringstream ss(line);
        std::string tmp;
        std::vector<std::string> v;
        while (ss >> tmp) {
            if (tmp[tmp.size() - 1] == ',')
                tmp.pop_back();
            v.push_back(tmp);
        }
        std::string name = v[0];
        int end = v[1].size();
        int weight = std::stoi(v[1].substr(1, end - 1));
        g[name].weight = weight;
        if (v.size() > 2) {
            std::vector<std::string> sub(v.begin() + 3,v.end());
            for (auto child: sub)
                g[name].progs.push_back(child);
        }
        //print("added/",name, g[name].weight, join(g[name].progs));
    }
    // do kahn
    std::unordered_map<std::string, int> indg;
    for (auto & [name,node] : g._map) {
        for (auto & child: node.progs)
            indg[child] += 1;
    }
    std::unordered_map<std::string, int>::iterator it = indg.begin(); 
    std::deque<std::string> degree0;
    for (auto & [name,_] : g._map) {
        if (indg.find(name) == indg.end())
            degree0.push_back(name);
    }
    assert (degree0.size() == 1);
    std::string p1 = degree0[0];
    std::vector<std::string> sorted;
    while (degree0.size()){
        std::string name = degree0.front();
        sorted.push_back(name);
        degree0.pop_front();
        for (auto & child : g[name].progs) { // childname
            indg[child]--;
            if (!indg[child])
                degree0.push_back(child);
        }
    }
    std::vector<std::string> reversed(sorted.rbegin(), sorted.rend());
    // now we got a reversedly topo-sorted tree
    std::unordered_map<std::string,int> SUMS;
    std::string special;
    int p2;
    for (auto name: reversed) {
        auto children = g[name].progs;
        // print("name/",name,"- progs/",join(children));
        std::unordered_map<int,int> FREQ;
        for (auto child: children)
            FREQ[SUMS[child]]++;
        assert (FREQ.size() < 3);
        int N = FREQ.size();
        int freqmax = -1, freqmin = INT_MAX, many, only;
        if (N == 2) {
            for (auto & [k,v] : FREQ) {
                if (freqmax < v) {
                    freqmax = v;
                    many = k;
                }
                if (freqmin > v) {
                    freqmin = v;
                    only = k;
                }
            }
        }
        int update = 0;
        std::string thewrongone;
        for (auto child: children) {
            update += SUMS[child];
            int accumulated = SUMS[child];
            if (N == 2 && accumulated ^ many) {
                thewrongone = child;
                break ;
            }
        }
        if (thewrongone != "") {
            int correction = std::abs(many - only);
            p2 = g[thewrongone].weight - correction;
            break ;
        }
        SUMS[name] = update + g[name].weight;
    }
    print("part 1:", p1);
    print("part 2:", p2);
}

